### DISCUSSION TITLE
How to access the positions direction from exchange?

### DISCUSSION THREAD

**Discussion by graceyangfan** [nautechsystems/nautilus_trader#609] [Q&A] [answered]
I am running a live code on binance,I found that when I  manually close all order on binance website,My strategy cannot recevied a feedback,it still think it has a position.I find that the strategy not directly query position from exhcange, where the code did not add position infomation to the account.
`https://github.com/nautechsystems/nautilus_trader/blob/82b8dfcb677bdb3a080970d3cb5face82e404685/nautilus_trader/adapters/binance/futures/execution.py#L265`
how your guys deal with this problem

**Comment by cjdsellers** [comment, marked as answer]
Hi @graceyangfan just so I understand the context better, when you say that it still thinks it has a position - was this on start-up or shutdown? Also, are you using a Redis backed cache, or in-memory (either should work).

On startup the `LiveExecutionEngine` will call `reconcile_state` on every adapter. All specific implementations work slightly differently, however in the case of Binance Futures it will query for all open order and positions which are open according to the exchange, and **also** all orders and positions which Nautilus believes to be open based on its current state. This is because unfortunately Binance don't allow you to query closed orders unless you pass a specific symbol, which would involve 2000 round trips for Spot/Margin, and 200 for Futures. In any case with the above strategy nothing should escape the systems notice.

**Reply by graceyangfan** [reply]
@cjdsellers  I use them in memory，I receive position infomation in stratgy on_event with PositionEvent，but when I closed the position on binance website， the stratgy will keep that it still have a position for a long time，I have to restart the program.Do the cache only store PositionChange locally instead of query postions directionly from binance？

**Reply by cjdsellers** [reply]
When you close the position on the website, this should create a market order fill event, which will be picked up by Nautilus over the websocket and assigned to the position. The currently released `1.141.0` doesn't handle external orders for Binance Spot/Margin, however the current `develop` branch does. If you're trading on Binance Futures then external orders should be correctly handled.

Could you please share your Binance execution client configuration?

**Reply by graceyangfan** [reply]
```
config_node = TradingNodeConfig(
    trader_id="TESTER-001",
    log_level="INFO",
    exec_engine={
        "reconciliation_lookback_mins": 1440,
    },
    cache_database=CacheDatabaseConfig(type="in-memory"),
    data_clients={
        "BINANCE": BinanceDataClientConfig(
            api_key=BINANCE_FUTURES_API_KEY,
            api_secret=BINANCE_FUTURES_API_SECRET,
            account_type=BinanceAccountType.FUTURES_USDT,
            instrument_provider=InstrumentProviderConfig(load_all=True),
        ),
    },
    exec_clients={
        "BINANCE": BinanceExecClientConfig(
            api_key=BINANCE_FUTURES_API_KEY,
            api_secret=BINANCE_FUTURES_API_SECRET,
            account_type=BinanceAccountType.FUTURES_USDT,
            instrument_provider=InstrumentProviderConfig(load_all=True),
        ),
    },
    timeout_connection=5.0,
    timeout_reconciliation=5.0,
    timeout_portfolio=5.0,
    timeout_disconnection=5.0,
    timeout_post_stop=2.0,
)

node = TradingNode(config=config_node)

strat_config1 = MyMarketMakerConfig(
    instrument_id="ETHUSDT-PERP.BINANCE",
    bar_type="ETHUSDT-PERP.BINANCE-1-MINUTE-LAST-EXTERNAL",
    My_period=30,
    My_multiple=4,
    atr_multiple = 2.5,
    max_take_profit = 5,
    stop_profit = 13,
    trade_size=Decimal("0.01"),
    strategy_id = '001'
)

strategy1 = MyMarketMaker(config=strat_config1)


strat_config2 = MyMarketMakerConfig(
    instrument_id="BNBUSDT-PERP.BINANCE",
    bar_type="BNBUSDT-PERP.BINANCE-1-MINUTE-LAST-EXTERNAL",
    My_period=20,
    My_multiple=5,
    atr_multiple = 3,
    max_take_profit = 2.5,
    stop_profit = 8,
    trade_size=Decimal("0.01"),
    strategy_id = '002'
)

strategy2 = MyMarketMaker(config=strat_config2)


node.trader.add_strategy(strategy1)
node.trader.add_strategy(strategy2)

node.add_data_client_factory("BINANCE", BinanceLiveDataClientFactory)
node.add_exec_client_factory("BINANCE", BinanceLiveExecClientFactory)
node.build()



if __name__ == "__main__":
    try:
        node.start()
    finally:
        node.dispose()
```

**Reply by cjdsellers** [reply]
This config looks fine to me

**Reply by graceyangfan** [reply]
Here is the position logical
```
class  MyMarketMaker(TradingStrategy):


    def __init__(self, config:  MyMarketMakerConfig):
        super().__init__(config)

        self.position: Optional[Position] = None

    def on_bar(self, bar: Bar):

        self.log.info(repr(bar), LogColor.CYAN)
        if not self.indicators_initialized():
            self.log.info(
                f"Waiting for indicators to warm up " f"[{self.cache.bar_count(self.bar_type)}]...",
                color=LogColor.BLUE,
            )
            return 

        last: QuoteTick = self.cache.quote_tick(self.instrument_id)
        if last is None:
            self.log.info("No quotes yet...")
            return

        if not self.position: 
            self.create_buy_order(last)

        if not self.position: 
            self.create_sell_order(last)

        

    def on_event(self, event: Event):

        if isinstance(event, PositionEvent):
            self.position = self.cache.position(event.position_id)

```
when closed positions on websit, if not self.position: does not work

**Reply by cjdsellers** [reply]
@graceyangfan were you able to resolve this?

**Reply by graceyangfan** [reply]
@cjdsellers There still have some problems on feedback to nautilus_trader，but I have stop doing this，for algorithm trading there should no need for haman intervention.

**Reply by cjdsellers** [reply]
OK, understood. Human oversight and intervention is often necessary especially in the early stages, and things always go wrong eventually.
